src/t_hash.c

```diff
@@ -1111,6 +1526,279 @@ static void hrandfieldReplyWithListpack(writePreparedClient *wpc, unsigned int c
+        }
+        addReplyLongLong(c, result);
+    }
+
+    if (expired || updated) {
+        if (expired) {
+            replaceClientCommandVector(c, new_argc, new_argv);
+            /* We would like to reduce the number of hexpired events in case there are potential many expired fields. */
+            notifyKeyspaceEvent(NOTIFY_HASH, "hexpired", c->argv[1], c->db->id);
+        } else if (updated) {
+            /* Propagate as HPEXPIREAT millisecond-timestamp
+             * Only rewrite the command arg if not already HPEXPIREAT */
+            if (c->cmd->proc != hpexpireAtCommand) {
+                rewriteClientCommandArgument(c, 0, shared.hpexpireat);
+            }
+
+            /* Avoid creating a string object when it's the same as argv[2] parameter  */
+            if (basetime != 0 || unit == UNIT_SECONDS) {
+                robj *when_obj = createStringObjectFromLongLong(when);
+                rewriteClientCommandArgument(c, 2, when_obj);
+                decrRefCount(when_obj);
+            }
+            notifyKeyspaceEvent(NOTIFY_HASH, "hexpire", c->argv[1], c->db->id);
+        }
+        server.dirty += (expired + updated); // in case there was a change increment the dirty
+        signalModifiedKey(c, c->db, obj);
+        /* Delete the object in case it was left empty */
+        if (hashTypeLength(obj) == 0) {
+            dbDelete(c->db, c->argv[1]);
+            notifyKeyspaceEvent(NOTIFY_GENERIC, "del", c->argv[1], c->db->id);
+        }
+    }
+}
+
+void hexpireCommand(client *c) {
+    hexpireGenericCommand(c, commandTimeSnapshot(), UNIT_SECONDS);
+}
+
+void hexpireAtCommand(client *c) {
+    hexpireGenericCommand(c, 0, UNIT_SECONDS);
+}
+
+void hpexpireCommand(client *c) {
+    hexpireGenericCommand(c, commandTimeSnapshot(), UNIT_MILLISECONDS);
+}
+
+void hpexpireAtCommand(client *c) {
+    hexpireGenericCommand(c, 0, UNIT_MILLISECONDS);
+}
+
+/* High-Level Algorithm of HPERSIST Command:
+ *
+ * - Expects a key and a list of hash fields whose expiration metadata should be removed.
+ * - Validates that the number of provided fields matches the declared count.
+ *
+ * - For each specified field attempts to remove any existing expiration.
+ * - Replies to the client  with an array of integers, each representing the result of persistence for one field:
+ *   - 1 if the expiration was set.
+ *   - -1 if the field exists, but has no expiraiton time set.
+ *   - -2 if the field does not exist or the hash is empty.
+ *
+ * - If any expirations were removed:
+ *   - Marks the key as modified (for replication/AOF consistency).
+ *   - Emits a "hpersist" keyspace notification.
+ *
+ * Keyspace Notifications (if enabled):
+ * - "hpersist" — emitted once if any field had its expiration removed. */
+void hpersistCommand(client *c) {
+    int fields_index = 4, result = 0, changes = 0;
+    long long num_fields = 0;
+
+    if (getLongLongFromObjectOrReply(c, c->argv[fields_index - 1], &num_fields, NULL) != C_OK) return;
```

Comment:
we have some similar commands which include numkeys as one argument in the argument list , such as BLMPOP and ZMPOP, they use another different way to check the num_fields, such as 
if (getRangeLongFromObjectOrReply(c, c->argv[numkeys_idx], 1, LONG_MAX, &numkeys,
                                      "numkeys should be greater than 0") != C_OK)
You can reference this, and then you do not need to condition !num_fields

Reply:
sounds good!

Reply:
@hwware although I lkike your suggestion, I ended up still using my existing check. the reason is that in this case we aim for exact match and not range check. so tell me what you think.